cmd/kubeadm/app/cmd/join.go

```diff
@@ -419,6 +421,20 @@ func (j *joinData) Run() error {
 		fmt.Printf("[join] Running pre-flight checks before initializing the new control plane instance\n")
 		preflight.RunInitMasterChecks(utilsexec.New(), initCfg, j.ignorePreflightErrors)
 
+		fmt.Printf("[join] Pulling control-plane images\n")
+		externalcfg := &kubeadmapiv1beta1.InitConfiguration{}
+		kubeadmscheme.Scheme.Default(externalcfg)
+		var cfgPath string
+		internalcfg, err := configutil.ConfigFileAndDefaultsToInternalConfig(cfgPath, externalcfg)
+		kubeadmutil.CheckErr(err)
+		containerRuntime, err := utilruntime.NewContainerRuntime(utilsexec.New(), internalcfg.GetCRISocket())
+		kubeadmutil.CheckErr(err)
+		imagesPull := NewImagesPull(containerRuntime, images.GetAllImages(&internalcfg.ClusterConfiguration))
```

Comment:
this call needs a package.
also should pass `&internalcfg.ClusterConfiguration` here.

the ClusterConfiguration is already know in this function. look at the call `fetchInitConfigurationFromJoinConfiguration`.

use `initCfg.ClusterConfiguration`

Reply:
Yes, we need to use the `ClusterConfiguration` that got downloaded from the cluster that we are intending to join, so we take into account specific registry locations and if the hyperkube image is used. Otherwise we would be downloading the default set of images that might not match the real images that will be used afterwards.

Reply:
thx for feedbacks updating :robot: